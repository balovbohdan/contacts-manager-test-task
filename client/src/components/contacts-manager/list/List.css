.main {

}

.loader {
    width: 100%;
}

.loader_img {
    width: 40px;
    height: 40px;

    margin: 0 auto;

    background-size: contain;
    background-repeat: no-repeat;
    background-position: center center;
    background-image: url(~@img/loader.svg);
}