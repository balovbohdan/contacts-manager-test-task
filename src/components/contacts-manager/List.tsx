import * as React from 'react';

import {T} from '@lib/contacts/contact';

import {Contact} from './Contact';
import css from './List.css';

type Props = {
    contacts:T.Contact[];
};

type ItemsProps = {
    contacts:T.Contact[];
};

export const List = ({contacts}:Props) =>
    <div className={css.main}>
        <Items contacts={contacts}/>
    </div>;

const Items = ({contacts}:ItemsProps) => {
    const mapper = ({id, img, name, phone}) => {
        id = id || name + phone;

        return <Contact id={id} key={id} img={img} name={name} phone={phone}/>;
    };

    const items = contacts.map(mapper);

    return <>{items}</>;
};