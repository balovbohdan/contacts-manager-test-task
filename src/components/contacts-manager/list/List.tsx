import * as React from 'react';

import css from './List.css';
import * as T from './types';
import {Contact} from '../contact';
import {DataProvider} from './DataProvider';

export const List = ({contacts, toggleContactWindow}:T.Props) =>
    <div className={css.main}>
        <DataProvider>{null}</DataProvider>
        <Items
            contacts={contacts}
            toggleContactWindow={toggleContactWindow}/>
    </div>;

const Items = ({contacts, toggleContactWindow}:T.ItemsProps) => {
    const mapper = contact =>
        <Item
            key={contact.id}
            contact={contact}
            toggleContactWindow={toggleContactWindow}/>;

    const items = Object.values(contacts).map(mapper);

    return <>{items}</>;
};

const Item = ({contact, toggleContactWindow}:T.ItemProps) =>
    <Contact
        contact={contact}
        toggleContactWindow={toggleContactWindow}/>;