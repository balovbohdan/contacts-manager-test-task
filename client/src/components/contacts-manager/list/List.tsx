import * as React from 'react';
import InfiniteScroll from 'react-infinite-scroll-component';

import * as css from './List.css';
import * as T from './types';
import {Contact} from '../contact';
import {DataProvider} from './DataProvider';

export const List = (props:T.Props) => {
    const contactsQty = props.contacts.length;
    const fetchContacts = () => props.fetchContacts({ contactsQty });

    return (
        <DataProvider fetchContacts={fetchContacts}>
            <Scroller
                contactsQty={contactsQty}
                fetchMore={fetchContacts}
                hasMore={props.hasMoreContacts}>
                <Items
                    contacts={props.contacts}
                    toggleContactWindow={props.toggleContactWindow}/>
            </Scroller>
        </DataProvider>
    );
};

const Scroller = ({hasMore, fetchMore, children, contactsQty}:T.ScrollerProps) => {
    const height = window.innerHeight - 170;

    return (
        <InfiniteScroll
            height={height}
            next={fetchMore}
            hasMore={hasMore}
            loader={<Loader/>}
            dataLength={contactsQty}>
            {children}
        </InfiniteScroll>
    );
};

const Items = ({contacts, toggleContactWindow}:T.ItemsProps) => {
    const mapper = contact =>
        <Item
            key={contact.id}
            contact={contact}
            toggleContactWindow={toggleContactWindow}/>;

    const items = contacts.map(mapper);

    return <>{items}</>;
};

const Item = ({contact, toggleContactWindow}:T.ItemProps) =>
    <Contact
        contact={contact}
        toggleContactWindow={toggleContactWindow}/>;

const Loader = () =>
    <div className={css.loader}>
        <div className={css.loader_img}/>
    </div>;