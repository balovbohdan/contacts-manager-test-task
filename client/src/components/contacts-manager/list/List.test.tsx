import * as React from 'react';
import {mount, shallow} from 'enzyme';

import {List} from './List';

describe('Contacts list', () => {
    const propsDef = {
        hasMoreContacts: true,
        fetchContacts: jest.fn(),
        toggleContactWindow: jest.fn(),
        contacts: [{
            id: 1,
            name: 'name',
            img: '/user.svg',
            phone: '+380975142353'
        }]
    };

    it('Should match snapshot', () => {
        const props = { ...propsDef };
        const list = shallow(<List {...props}/>);

        expect(list).toMatchSnapshot();
    });

    it('Should call contacts fetcher function', () => {
        const props = {
            ...propsDef,
            fetchContacts: jest.fn()
        };

        mount(<List {...props}/>);
        expect(props.fetchContacts).toBeCalledTimes(1);
    });
});