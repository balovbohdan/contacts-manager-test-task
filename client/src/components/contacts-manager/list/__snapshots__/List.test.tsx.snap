// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contacts list Should match snapshot 1`] = `
<Component
  fetchContacts={[Function]}
>
  <Scroller
    contactsQty={1}
    fetchMore={[Function]}
    hasMore={true}
  >
    <Items
      contacts={
        Array [
          Object {
            "id": 1,
            "img": "/user.svg",
            "name": "name",
            "phone": "+380975142353",
          },
        ]
      }
      toggleContactWindow={[MockFunction]}
    />
  </Scroller>
</Component>
`;
